### ISSUE TITLE
TypeError: optionConstructors[code] is not a function

### ISSUE THREAD

**Issue by CliffS** [AlCalzone/node-coap-client#21]
This is happening when calling  `connect()` from `node-tradfri-client`.  Please let me know if you need further information.

The stack trace is:
```
TypeError: optionConstructors[code] is not a function
    at Function.parse (/home/cliff/src/git/house/node_modules/node-coap-client/build/Option.js:93:45)
    at Function.parse (/home/cliff/src/git/house/node_modules/node-coap-client/build/Message.js:108:44)
    at Function.onMessage (/home/cliff/src/git/house/node_modules/node-coap-client/build/CoapClient.js:444:43)
    at SocketWrapper.emit (events.js:159:13)
    at SocketWrapper.socket.on (/home/cliff/src/git/house/node_modules/node-coap-client/build/lib/SocketWrapper.js:12:18)
    at Socket.emit (events.js:159:13)
    at Socket.udp_onMessage (/home/cliff/src/git/house/node_modules/node-dtls-client/build/dtls.js:204:34)
    at Socket.emit (events.js:159:13)
    at UDP.onMessage (dgram.js:658:8)
```
At this point, `code` is `23`, `prevCode` is `14` and `delta` is `9`.

**Comment by CliffS** [comment]
Hmm... it seems that `connect()` succeeds and **then** I get the error.  The gateway is running v1.2.42, if that means anything to you.

**Comment by AlCalzone** [comment]
There's nothing in the CoAP specs about an option with code 23. Do you have a possibility to output the raw buffer `buf` inside `Option.parse` or even better in the calling `Message.parse`?

**Comment by CliffS** [comment]
The buffer at the point where `code` is 23 contains (in hex):

```
910eff7b2239303031223a224556455259444159222c2239303032223a313530383235
303135392c2239303638223a312c2239303033223a3230303436382c2239303537223a
302c223135303133223a5b7b2235383530223a312c2235383531223a3230332c223537
3037223a353432372c2235373038223a34323539362c2235373039223a33303031352c
2235373130223a32363837302c2235373131223a302c2235373036223a226631653062
35222c2239303033223a36353533397d2c7b2235383530223a312c2235383531223a32
30332c2235373037223a353432372c2235373038223a34323539362c2235373039223a
33303031352c2235373130223a32363837302c2235373131223a302c2235373036223a
22663165306235222c2239303033223a36353534307d2c7b2235383530223a312c2235
383531223a3230332c2235373037223a353432372c2235373038223a34323539362c22
35373039223a33303031352c2235373130223a32363837302c2235373131223a302c22
35373036223a22663165306235222c2239303033223a36353534317d2c7b2235383530
223a312c2235383531223a3230332c2235373037223a353432372c2235373038223a34
323539362c2235373039223a33303031352c2235373130223a32363837302c22353731
31223a302c2235373036223a22663165306235222c2239303033223a36353534327d2c
7b2235383530223a312c2235383531223a3230332c2235373037223a353432372c2235
373038223a34323539362c2235373039223a33303031352c2235373130223a32363837
302c2235373131223a302c2235373036223a22663165306235222c2239303033223a36
353534337d2c7b2235383530223a312c2235383531223a3230332c2235373037223a35
3432372c2235373038223a34323539362c2235373039223a33303031352c2235373130
223a32363837302c2235373131223a302c2235373036223a22663165306235222c2239
303033223a36353534347d2c7b2235383530223a312c2235383531223a3230332c2235
373037223a353432372c2235373038223a34323539362c2235373039223a3330303135
2c2235373130223a32363837302c2235373131223a302c2235373036223a2266316530
6235222c2239303033223a36353534357d2c7b2235383530223a312c2235383531223a
3230332c2235373037223a353432372c2235373038223a34323539362c223537303922
3a33303031352c2235373130223a32363837302c2235373131223a302c223537303622
3a22663165306235222c2239303033223a36353534367d2c7b2235383530223a312c22
35383531223a3230332c2235373037223a353432372c2235373038223a34323539362c
2235373039223a3330303135
```
This was output by inserting the following code just before the `return` statement.

```javascript
        if (code == 23) {
          console.log(buf.toString('hex'));
        }
```
I've attached it as a text file too.

[buf.txt](https://github.com/AlCalzone/node-coap-client/files/1540042/buf.txt)

**Comment by AlCalzone** [comment]
Thanks! I cannot promise when I'll be able to check it out. Was that the buffer inside `Option.parse` or the one inside `Message.parse`?

**Comment by CliffS** [comment]
> Was that the buffer inside `Option.parse` or the one inside `Message.parse`?

That was from `Option.parse`.  From `Message.parse`, I assume you want the raw buffer on entry into the routine?  If so, the output is:

```
64450025018fccf460613223093a80910eff7b2239303031223a224556455259444159
222c2239303032223a313530383235303135392c2239303638223a312c223930303322
3a3230303436382c2239303537223a302c223135303133223a5b7b2235383530223a31
2c2235383531223a3230332c2235373037223a353432372c2235373038223a34323539
362c2235373039223a33303031352c2235373130223a32363837302c2235373131223a
302c2235373036223a22663165306235222c2239303033223a36353533397d2c7b2235
383530223a312c2235383531223a3230332c2235373037223a353432372c2235373038
223a34323539362c2235373039223a33303031352c2235373130223a32363837302c22
35373131223a302c2235373036223a22663165306235222c2239303033223a36353534
307d2c7b2235383530223a312c2235383531223a3230332c2235373037223a35343237
2c2235373038223a34323539362c2235373039223a33303031352c2235373130223a32
363837302c2235373131223a302c2235373036223a22663165306235222c2239303033
223a36353534317d2c7b2235383530223a312c2235383531223a3230332c2235373037
223a353432372c2235373038223a34323539362c2235373039223a33303031352c2235
373130223a32363837302c2235373131223a302c2235373036223a2266316530623522
2c2239303033223a36353534327d2c7b2235383530223a312c2235383531223a323033
2c2235373037223a353432372c2235373038223a34323539362c2235373039223a3330
3031352c2235373130223a32363837302c2235373131223a302c2235373036223a2266
3165306235222c2239303033223a36353534337d2c7b2235383530223a312c22353835
31223a3230332c2235373037223a353432372c2235373038223a34323539362c223537
3039223a33303031352c2235373130223a32363837302c2235373131223a302c223537
3036223a22663165306235222c2239303033223a36353534347d2c7b2235383530223a
312c2235383531223a3230332c2235373037223a353432372c2235373038223a343235
39362c2235373039223a33303031352c2235373130223a32363837302c223537313122
3a302c2235373036223a22663165306235222c2239303033223a36353534357d2c7b22
35383530223a312c2235383531223a3230332c2235373037223a353432372c22353730
38223a34323539362c2235373039223a33303031352c2235373130223a32363837302c
2235373131223a302c2235373036223a22663165306235222c2239303033223a363535
34367d2c7b2235383530223a312c2235383531223a3230332c2235373037223a353432
372c2235373038223a34323539362c2235373039223a3330303135
```
[message.txt](https://github.com/AlCalzone/node-coap-client/files/1542308/message.txt)

**Comment by AlCalzone** [comment]
For some reason, your gateway uses options mine doesn't (its on the same version). Apparently code 23 was defined in https://tools.ietf.org/html/rfc7959 "block-wise transfer". I'll have to read that spec and implement it.

**Comment by CliffS** [comment]
> I'll have to read that spec and implement it.

No rush.  Any time today is fine. :)

**Comment by AlCalzone** [comment]
> Any time today is fine. :)

Alright, I'll give it a try after the christmas party at work. I hear a couple of beers help creating bug-free code.

**Comment by CliffS** [comment]
Not nagging as I appreciate you're doing all this for free but do you have a timescale when you will be able to look at it again.  I'd offer to help but I don't understand either typescript or CoAP.

**Comment by AlCalzone** [comment]
Probably tomorrow. Last week was insanely busy.

**Comment by AlCalzone** [comment]
I might have a new alpha version for you to test tomorrow. When the lib no longer crashes during parsing of those blockwise messages, I will need another log to test assembling multiple partial messages into one.

**Comment by CliffS** [comment]
That's great.  I can set up to test from your working branch if that makes it easier.

**Comment by AlCalzone** [comment]
I'll give you the commit hash as I can't be sure the branch is always in a working state.

**Comment by AlCalzone** [comment]
I realized I can just force my own gateway to respond with unnecessarily small messages. No need for you to test half-baked versions.

**Comment by AlCalzone** [comment]
Commit b041be3 should fix your issue. Please test, so I can merge.

**Comment by CliffS** [comment]
I will test tomorrow morning.  Many thanks.

**Comment by CliffS** [comment]
Hi.  Sorry it has taken me so long to get to this.  It seems to work perfectly now.

**Comment by AlCalzone** [comment]
No problem at all! Thanks for testing and happy holidays!

**Comment by CliffS** [comment]
Thank you for sorting this out so quickly.  Always amazed the great service available for free, open-source software.

Have a great Christmas.

**Comment by AlCalzone** [comment]
Fixes included in v0.5.0 (on NPM aswell)

**Comment by CliffS** [comment]
> Fixes included in v0.5.0 (on NPM aswell)

Please don't forget to update the dependency in node-tradfri-client.

Thanks.

**Comment by AlCalzone** [comment]
Good catch!
